lib/utils.ts

```diff
@@ -164,3 +163,11 @@ export function isJsonString(str: string) {
   }
   return true;
 }
+
+export function openAiCost(messages: ChatGPTMessage[]): number {
+  const encoded = tokenizer.encodeChat(messages as ChatMessage[], "gpt-4");
+  const nTokens = encoded.length;
+  // For the 8k context model
+  const gpt4CostPerToken = 0.03 / 1000;
+  return nTokens * gpt4CostPerToken;
+}
```

Comment:
This doesn't take into account the difference in cost between input and output tokens. Also, since we only track at the end, we miss out usage if there are errors.

To simplify this we could either:
1. Use the openai billing api on the usage page instead of storing ourselves (https://api.openai.com/v1/dashboard/billing/usage?start_date=2023-06-01&end_date=2023-06-05)
2. Use the `usage` field from the openai response to see the cost of input & output tokens, as calculated by them

Reply:
The usage field option doesn't work unfortunately as it's not returned when stream=true https://community.openai.com/t/openai-api-get-usage-tokens-in-response-when-set-stream-true/141866